### PR TITLE
Metadata: escape display name in email addresses

### PR DESCRIPTION
**Pull request by robsdedude**
This follows the way setuptools handles email addresses from `pyproject.toml` files.

See [here](https://github.com/pypa/setuptools/blob/v68.2.2/setuptools/config/_apply_pyprojecttoml.py#L192-L193) where it utilizes the std lib package `email`, which in turn escapes email addresses [here](https://github.com/python/cpython/blob/v3.12.0/Lib/email/headerregistry.py#L89-L90) and then [here](https://github.com/python/cpython/blob/v3.12.0/Lib/email/_header_value_parser.py#L97).

I noticed this bug when I tried to upload a package to PyPI with something like this in my `pyproject.toml`:
```toml
[project]
# ...
authors = [
    {name = "FooBar, Inc.", email = "contact@foobar.com"}
]
# ...
```

PyPI rejected the package:
```
HTTPError: 400 Bad Request from https://test.pypi.org/legacy/
'FooBar, Inc. <contact@nfoobar.com>' is an invalid value for Author-email.
Error: Use a valid email address See
https://packaging.python.org/specifications/core-metadata for more
information.
```

As a workaround, I changed my pyproject.toml to
```toml
[project]
# ...
authors = [
    {email = "\"FooBar, Inc.\" <contact@foobar.com>"}
]
# ...
```

### CODE REVIEW

File: src/metadata.rs [hunk 1]
```diff
@@ -555,6 +555,23 @@ impl Metadata21 {
     }
 }
 
+/// Escape email addresses with display name if necessary
+/// according to RFC 822 Section 3.3. "specials".
+fn format_email_with_display_name(display_name: &str, email: &str) -> String {
```
Comment (messense) [inline src/metadata.rs:560]:
```suggestion
fn escape_email_with_display_name(display_name: &str, email: &str) -> String {
```

File: src/metadata.rs [hunk 2]
```diff
@@ -226,7 +226,7 @@ impl Metadata21 {
                 for author in authors {
                     match (&author.name, &author.email) {
                         (Some(name), Some(email)) => {
-                            emails.push(format!("{name} <{email}>"));
+                            emails.push(format_email_with_display_name(name, email));
```
Comment (messense) [inline src/metadata.rs:229]:
```suggestion
                            emails.push(escape_email_with_display_name(name, email));
```

File: src/metadata.rs [hunk 3]
```diff
@@ -251,7 +251,7 @@ impl Metadata21 {
                 for maintainer in maintainers {
                     match (&maintainer.name, &maintainer.email) {
                         (Some(name), Some(email)) => {
-                            emails.push(format!("{name} <{email}>"));
+                            emails.push(format_email_with_display_name(name, email));
```
Comment (messense) [inline src/metadata.rs:254]:
```suggestion
                            emails.push(escape_email_with_display_name(name, email));
```

**Review by messense — COMMENTED** [review]
Thanks! Can you also add a test case for the new function?

**Review by robsdedude — COMMENTED** [review]
Rename function

**Comment by robsdedude** [comment]
@messense tests added.

**Review by messense — APPROVED** [review]
LGTM, thanks!

**Pull request merged**